### PR TITLE
Drop unused columns after join on/using

### PR DESCRIPTION
**Pull request by vdimir**
### Changelog category (leave one):

- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by vdimir** [comment]
Difference (running added test on current version):

<details>
<summary>Details</summary>



```
02514_analyzer_drop_join_on:                                            [ FAIL ] - result differs with reference: 
--- /home/ubuntu/ClickHouseA/tests/queries/0_stateless/02514_analyzer_drop_join_on.reference    2022-12-27 11:16:11.639907351 +0000
+++ /home/ubuntu/ClickHouseA/tests/queries/0_stateless/02514_analyzer_drop_join_on.stdout       2022-12-27 11:16:36.235829022 +0000
@@ -7,17 +7,23 @@
     Header: default.a.a2_4 String
       Join (JOIN FillRightFirst)
       Header: default.a.a2_4 String
+              default.a.a1_1 UInt64
+              default.b.b1_0 UInt64
               default.c.c1_2 UInt64
               default.d.d1_3 UInt64
         Expression ((JOIN actions + DROP unused columns after JOIN))
         Header: default.a.a2_4 String
+                default.a.a1_1 UInt64
+                default.b.b1_0 UInt64
                 default.c.c1_2 UInt64
           Join (JOIN FillRightFirst)
           Header: default.a.a2_4 String
+                  default.a.a1_1 UInt64
                   default.b.b1_0 UInt64
                   default.c.c1_2 UInt64
             Expression ((JOIN actions + DROP unused columns after JOIN))
             Header: default.a.a2_4 String
+                    default.a.a1_1 UInt64
                     default.b.b1_0 UInt64
               Join (JOIN FillRightFirst)
               Header: default.a.a2_4 String
@@ -47,18 +53,24 @@
   Join (JOIN FillRightFirst)
   Header: default.a.k_2 UInt64
           default.a.a2_0 String
+          default.b.k_3 UInt64
+          default.c.k_4 UInt64
           default.d.d2_1 String
           default.d.k_5 UInt64
     Expression (DROP unused columns after JOIN)
     Header: default.a.k_2 UInt64
             default.a.a2_0 String
+            default.b.k_3 UInt64
+            default.c.k_4 UInt64
       Join (JOIN FillRightFirst)
       Header: default.a.k_2 UInt64
               default.a.a2_0 String
+              default.b.k_3 UInt64
               default.c.k_4 UInt64
         Expression (DROP unused columns after JOIN)
         Header: default.a.k_2 UInt64
                 default.a.a2_0 String
+                default.b.k_3 UInt64
           Join (JOIN FillRightFirst)
           Header: default.a.k_2 UInt64
                   default.a.a2_0 String
@@ -93,23 +105,29 @@
             b.bx_0 String
       Join (JOIN FillRightFirst)
       Header: default.a.a2_6 String
+              default.a.a1_2 UInt64
               b.bx_0 String
+              b.b1_1 UInt64
               default.c.c2_5 String
               default.c.c1_3 UInt64
               d.d1_4 UInt64
         Filter (( + (JOIN actions + DROP unused columns after JOIN)))
         Header: default.a.a2_6 String
+                default.a.a1_2 UInt64
                 b.bx_0 String
+                b.b1_1 UInt64
                 default.c.c2_5 String
                 default.c.c1_3 UInt64
           Join (JOIN FillRightFirst)
           Header: default.a.a2_6 String
+                  default.a.a1_2 UInt64
                   b.bx_0 String
                   b.b1_1 UInt64
                   default.c.c2_5 String
                   default.c.c1_3 UInt64
             Expression ((JOIN actions + DROP unused columns after JOIN))
             Header: default.a.a2_6 String
+                    default.a.a1_2 UInt64
                     b.bx_0 String
                     b.b1_1 UInt64
               Join (JOIN FillRightFirst)
```



</details>

File: src/Planner/CollectColumnIndetifiers.h [hunk 1]
```diff
@@ -0,0 +1,22 @@
+#pragma once
```
Comment (kitaisreal) [inline src/Planner/CollectColumnIndetifiers.h:1]: Need to rename this file into `CollectColumnIdentifiers.h` from `CollectColumnIndetifiers.h`.

File: src/Planner/CollectColumnIndetifiers.h [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+#include <Analyzer/IQueryTreeNode.h>
+#include <Planner/Planner.h>
+#include <Planner/TableExpressionData.h>
+
+namespace DB
+{
+
+using ColumnIdentifierSet = std::unordered_set<ColumnIdentifier>;
+
+/// Collect all top level column identifiers from query tree node.
```
Comment (kitaisreal) [inline src/Planner/CollectColumnIndetifiers.h:12]: Need to add in comment precondition that table expression data must be already collected in planner context.

File: src/Planner/PlannerJoinTree.h [hunk 3]
```diff
@@ -11,10 +11,13 @@
 namespace DB
 {
 
+using ColumnIdentifierSet = std::unordered_set<ColumnIdentifier>;
```
Comment (kitaisreal) [inline src/Planner/PlannerJoinTree.h:14]: Need to move this into `TableExpressionData` and reuse in `CollectColumnIdentifiers`.

File: tests/queries/0_stateless/02514_analyzer_drop_join_on.sql [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+CREATE TABLE c (k UInt64, c1 UInt64, c2 String) ENGINE = Memory;
+INSERT INTO c VALUES (1, 1, 'a'), (2, 2, 'b'), (3, 3, 'c');
+
+CREATE TABLE d (k UInt64, d1 UInt64, d2 String) ENGINE = Memory;
+INSERT INTO d VALUES (1, 1, 'a'), (2, 2, 'b'), (3, 3, 'c');
+
+SET allow_experimental_analyzer = 1;
+
+EXPLAIN PLAN header = 1
```
Comment (kitaisreal) [inline tests/queries/0_stateless/02514_analyzer_drop_join_on.sql:20]: Consider to enable `echoOn` so test will be easier to verify.

**Review by kitaisreal — CHANGES_REQUESTED** [review]
Looks good. Need to fix minor details and merge.

**Review by kitaisreal — APPROVED** [review]

**Review by kitaisreal — APPROVED** [review]

**Pull request merged**